File: as/scan.c
/*
 * scan.c
 * Copyright (C) 2013-2015  Mikael Pettersson
 *
 * This file is part of pdp10-tools.
 *
 * pdp10-tools is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * pdp10-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pdp10-tools.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <errno.h>
#include <limits.h>	/* XXX: for UCHAR_MAX, deleteme */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "scan.h"
#include "token.h"

void scan_init(struct scan_state *scan_state, const char *progname)
{
    scan_state->progname = progname;
    scan_state->filename = "<stdin>";
    scan_state->linenr = 1;
}

int scan_open(struct scan_state *scan_state, const char *filename)
{
    if (filename[0] == '-' && filename[1] == '-' && filename[2] == '\0') {
	scan_state->filename = "<stdin>";
	filename = "/dev/stdin";
    } else
	scan_state->filename = filename;

    if (freopen(filename, "r", stdin) == NULL) {
	fprintf(stderr, "%s: Error opening %s: %s\n", scan_state->progname, filename, strerror(errno));
	return -1;
    }

    return 0;
}

static void scan_ungetc(struct scan_state *scan_state, int ch)
{
    if (ch != EOF && ungetc(ch, stdin) == EOF)
	fprintf(stderr, "%s: %s line %u: ungetc %d failed: %s\n",
		scan_state->progname, scan_state->filename, scan_state->linenr, ch, strerror(errno));
}

static int scan_getchar(void)
{
    return fgetc(stdin);
}

static void badchar(struct scan_state *scan_state, int ch, const char *context)
{
    char buf[7];

    if (ch == EOF) {
	buf[0] = '<';
	buf[1] = 'E';
	buf[2] = 'O';
	buf[3] = 'F';
	buf[4] = '>';
	buf[5] = '\0';
    } else if (' ' <= ch && ch <= '~') {
	buf[0] = '\'';
	buf[1] = ch;
	buf[2] = '\'';
	buf[3] = '\0';
    } else {
	buf[0] = '\'';
	buf[1] = '\\';
	buf[2] = '0' + ((ch >> 6) & 3);
	buf[3] = '0' + ((ch >> 3) & 7);
	buf[4] = '0' + (ch & 7);
	buf[5] = '\'';
	buf[6] = '\0';
    }

    fprintf(stderr, "%s: %s line %u: invalid character %s%s\n",
	    scan_state->progname, scan_state->filename, scan_state->linenr, buf, context);
}

static unsigned int get_chval(int ch)
{
    if ('0' <= ch && ch <= '9')
	return ch - '0';
    if ('A' <= ch && ch <= 'F')
	return ch - ('A' - 10);
    if ('a' <= ch && ch <= 'f')
	return ch - ('a' - 10);
    return -1U;
}

static int is_octal_digit(int ch)
{
    return ch >= '0' && ch <= '7';
}

static int do_escape(struct scan_state *scan_state)
{
    int ch;

    ch = scan_getchar();
    switch (ch) {
    case 'n':
	return '\n';
    case 't':
	return '\t';
    case 'f':
	return '\f';
    case 'r':
	return '\r';
    case 'b':
	return '\b';
    case '\\':
    case '\'':
    case '"':
	return ch;
    default:
	if (is_octal_digit(ch)) {
	    unsigned int val = ch - '0';
	    ch = scan_getchar();
	    if (is_octal_digit(ch)) {
		val = val * 8 + (ch - '0');
		ch = scan_getchar();
		if (is_octal_digit(ch))
		    val = val * 8 + (ch - '0');
		else
		    scan_ungetc(scan_state, ch);
	    } else
		scan_ungetc(scan_state, ch);
	    /* XXX: this should be PDP10_UINT9_MAX, but our string elements are still char not pdp10_uint9_t for now */
	    if (val > UCHAR_MAX) {
		fprintf(stderr, "%s: %s line %u: out of range character escape value %#x\n",
			scan_state->progname, scan_state->filename, scan_state->linenr, val);
		return EOF;
	    }
	    return val & UCHAR_MAX;
	}
	break;
    }
    badchar(scan_state, ch, "in \\ character escape");
    if (ch == '\n')
	++scan_state->linenr;
    return EOF;
}

/* XXX: string literals should be sequences of pdp10_uint9_t, not sequences of char */

static enum token do_string(struct scan_state *scan_state, union token_attribute *token_attr)
{
    char charbuf[4096];	/* 4095 char + NUL, XXX: make it dynamic */
    unsigned int len;
    char *text;
    int ch;

    len = 0;
    for (;;) {
	ch = scan_getchar();
	switch (ch) {
	case '"':
	    text = malloc(len + 1);
	    if (!text) {
		fprintf(stderr, "%s: %s line %u: malloc(%u) failed: %s\n",
			scan_state->progname, scan_state->filename, scan_state->linenr, len + 1, strerror(errno));
		return T_ERROR;
	    }
	    strcpy(text, charbuf);
	    token_attr->text = text;
	    return T_STRING;
	case '\\':
	    ch = do_escape(scan_state);
	    if (ch == EOF)
		return T_ERROR;
	    break;
	case EOF:
	case '\n':
	    badchar(scan_state, ch, "in string literal");
	    if (ch == '\n')
		++scan_state->linenr;
	    return T_ERROR;
	default:
	    break;
	}
	if (len >= sizeof charbuf - 1) {
	    fprintf(stderr, "%s: %s line %u: too long string literal\n",
		    scan_state->progname, scan_state->filename, scan_state->linenr);
	    return T_ERROR;
	}
	charbuf[len] = ch;
	++len;
    }
}

static int is_symbol_internal_char(int ch)
{
    return
	('A' <= ch && ch <= 'Z')
	|| ('a' <= ch && ch <= 'z')
	|| ('0' <= ch && ch <= '9')
	|| ch == '_'
	|| ch == '$'
	|| ch == '.';
}

static enum token do_symbol(struct scan_state *scan_state, union token_attribute *token_attr, int ch)
{
    char charbuf[128];	/* 127 chars + NUL, XXX: make it dynamic */
    unsigned int len;
    char *text;

    len = 0;
    do {
	if (len >= sizeof charbuf - 1) {
	    fprintf(stderr, "%s: %s line %u: too long symbol\n",
		    scan_state->progname, scan_state->filename, scan_state->linenr);
	    return T_ERROR;
	}
	charbuf[len] = ch;
	++len;
	ch = scan_getchar();
    } while (is_symbol_internal_char(ch));
    charbuf[len] = '\0';
    scan_ungetc(scan_state, ch);
    
    if (charbuf[0] == '.') {
	enum token low, high;

	if (charbuf[1] == '\0')
	    return T_DOT;

	/* see token.def, reserved symbols occupy tokens [0,T_SYMBOL[ */
	low = 0;
	high = T_SYMBOL;

	while (low < high) {
	    enum token middle;
	    int cmp;

	    middle = (low + high) / 2;
	    cmp = strcmp(charbuf, token_info[middle].print_name);

	    if (cmp < 0)
		high = middle;
	    else if (cmp > 0)
		low = middle + 1;
	    else
		return middle;
	}
    }

    text = malloc(len + 1);
    if (!text) {
	fprintf(stderr, "%s: %s line %u: malloc(%u) failed: %s\n",
		scan_state->progname, scan_state->filename, scan_state->linenr, len + 1, strerror(errno));
	return T_ERROR;
    }
    strcpy(text, charbuf);
    token_attr->text = text;
    return T_SYMBOL;
}

static enum token do_number(struct scan_state *scan_state, union token_attribute *token_attr, int ch)
{
    unsigned int base, chval;
    pdp10_uint36_t numval;

    base = (ch == '0') ? 8 : 10;
    numval = ch - '0';

    /* handle 0x<first hexdig> */
    ch = scan_getchar();
    if (base == 8 && (ch == 'x' || ch == 'X')) {
	base = 16;
	/* must have at least one hex digit after 0x */
	ch = scan_getchar();
	chval = get_chval(ch);
	if (chval >= 16) {
	    badchar(scan_state, ch, " after 0x in hexadecimal literal");
	    return T_ERROR;
	}
	numval = chval;
	ch = scan_getchar();
    }

    /* the number is non-empty, consume and accumulate trailing
       characters as long as they are valid in the base */
    for (;;) {
	chval = get_chval(ch);
	if (chval >= base)
	    break;
	numval = numval * base + chval;	/* XXX: check for overflow */
	ch = scan_getchar();
    }

    /* XXX: check for <decimal>{b,f} which is a local label reference */

    /* plain integer literal */
    scan_ungetc(scan_state, ch);
    token_attr->uint = numval;
    return T_UINTEGER;
}

static int do_line_comment(struct scan_state *scan_state)
{
    for (;;) {
	int ch = scan_getchar();
	switch (ch) {
	case '\n':
	    ++scan_state->linenr;
	    return T_NEWLINE;
	case EOF:
	    badchar(scan_state, ch, "in line comment");
	    return T_ERROR;
	default:
	    continue;
	}
    }
}

enum token scan_token(struct scan_state *scan_state, union token_attribute *token_attr)
{
    int ch;

    ch = scan_getchar();

    for (;; ch = scan_getchar()) {
	switch (ch) {
	case ' ':
	case '\t':
	case '\r':
	case '\f':
	    continue;
	case '\n':
	    ++scan_state->linenr;
	    return T_NEWLINE;
	case '#':
	    return do_line_comment(scan_state);
	case EOF:
	    return T_EOF;
	case '@':
	    return T_AT;
	case ':':
	    return T_COLON;
	case ',':
	    return T_COMMA;
	case '(':
	    return T_LPAREN;
	case ')':
	    return T_RPAREN;
	case '"':
	    return do_string(scan_state, token_attr);
	case '-':
	    return T_MINUS;
	case '.':
	    /* Dot may start a floating point literal, but tests show that
	       gcc always outputs floating point values as integer literals,
	       so we shouldn't have to support floating point literals at all.  */
	case '$':
	case '_':
	    return do_symbol(scan_state, token_attr, ch);
	default:
	    if ('0' <= ch && ch <= '9')	/* number or <decimal>{b,f} */
		return do_number(scan_state, token_attr, ch);
	    if (('A' <= ch && ch <= 'Z') ||
		('a' <= ch && ch <= 'z'))
		return do_symbol(scan_state, token_attr, ch);
	    break;
	}
	badchar(scan_state, ch, "");
	return T_ERROR;
    }
}

File: erlang/apps/as/src/scan.erl
%%% -*- erlang-indent-level: 2 -*-
%%%
%%% scanner for pdp10-elf as
%%% Copyright (C) 2013-2019  Mikael Pettersson
%%%
%%% This file is part of pdp10-tools.
%%%
%%% pdp10-tools is free software: you can redistribute it and/or modify
%%% it under the terms of the GNU General Public License as published by
%%% the Free Software Foundation, either version 3 of the License, or
%%% (at your option) any later version.
%%%
%%% pdp10-tools is distributed in the hope that it will be useful,
%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%%% GNU General Public License for more details.
%%%
%%% You should have received a copy of the GNU General Public License
%%% along with pdp10-tools.  If not, see <http://www.gnu.org/licenses/>.

-module(scan).

-export([ token/1
        , format_error/1
        , stdin/0
        , fopen/1
        , fclose/1
        ]).

-include("token.hrl").

-type scan_state() :: {scan_state, reference()}.
-type location() :: {Filename :: string(), LineNr :: pos_integer()}.

-export_type([scan_state/0, location/0]).

%% Scan State ------------------------------------------------------------------

%% The scanner state records the I/O handle, implements a one-character
%% pushback buffer, and maintains the current line number.
%% TODO: maintain column number too?
-record(scan_state,
        { filename      :: string()
        , iodev         :: file:fd() | standard_io
        , ungetc        :: [] | byte()
        , linenr        :: pos_integer()
        }).

-spec fclose(scan_state()) -> ok.
fclose(Handle) ->
  ScanState = #scan_state{} = get(Handle),
  case ScanState#scan_state.iodev of
    standard_io -> ok;
    IoDev -> file:close(IoDev)
  end,
  erase(Handle),
  ok.

-spec fgetc(scan_state()) -> {ok, byte()} | eof | {error, {module(), term()}}.
fgetc(Handle) ->
  ScanState = #scan_state{} = get(Handle),
  case ScanState#scan_state.ungetc of
    [] ->
      case file:read(ScanState#scan_state.iodev, 1) of
        {ok, [Byte]} ->
          case Byte of
            $\n ->
              put(Handle, ScanState#scan_state{linenr = ScanState#scan_state.linenr + 1}),
              {ok, $\n};
            _ ->
              {ok, Byte}
          end;
        eof ->
          eof;
        {error, Reason} ->
          {error, {file, Reason}}
      end;
    Ch ->
      put(Handle, ScanState#scan_state{ungetc = []}),
      {ok, Ch}
  end.

-spec fopen(string()) -> {ok, scan_state()} | {error, {module(), term()}}.
fopen(Filename) ->
  case file:open(Filename, [raw, read, read_ahead]) of
    {ok, IoDev} -> do_fopen(Filename, IoDev);
    {error, Reason} -> {error, {file, Reason}}
  end.

-spec stdin() -> {ok, scan_state()}.
stdin() ->
  do_fopen(_Filename = "<stdin>", _IoDev = standard_io).

do_fopen(Filename, IoDev) ->
  ScanState = #scan_state{ filename = Filename
                         , iodev = IoDev
                         , ungetc = []
                         , linenr = 1
                         },
  Handle = {scan_state, make_ref()},
  put(Handle, ScanState),
  {ok, Handle}.

-spec ungetc(byte(), scan_state()) -> ok | {error, {module(), term()}}.
ungetc(Ch, Handle) ->
  ScanState = #scan_state{} = get(Handle),
  case ScanState#scan_state.ungetc of
    [] ->
      put(Handle, ScanState#scan_state{ungetc = Ch}),
      ok;
    _ ->
      {error, {?MODULE, ungetc}}
  end.

-spec location(scan_state()) -> {ok, location()}.
location(Handle) ->
  ScanState = #scan_state{} = get(Handle),
  Location = {ScanState#scan_state.filename, ScanState#scan_state.linenr},
  {ok, Location}.

%% Scanner ---------------------------------------------------------------------

-spec token(scan_state())
      -> {ok, {location(), token()}} | {error, {module(), term()}}.
token(ScanState) ->
  %% TODO: optimize
  {ok, Location} = location(ScanState),
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> {ok, {Location, ?T_EOF}};
    {ok, Ch} ->
      case Ch of
        $\s -> token(ScanState);
        $\t -> token(ScanState);
        $\r -> token(ScanState);
        $\f -> token(ScanState);
        $\n -> {ok, {Location, ?T_NEWLINE}};
        $#  -> do_line_comment(ScanState);
        $@  -> {ok, {Location, ?T_AT}};
        $:  -> {ok, {Location, ?T_COLON}};
        $;  -> {ok, {Location, ?T_NEWLINE}};
        $,  -> {ok, {Location, ?T_COMMA}};
        $(  -> {ok, {Location, ?T_LPAREN}};
        $)  -> {ok, {Location, ?T_RPAREN}};
        $/  -> do_slash(ScanState);
        $\" -> do_string(ScanState, Location, []);
        $-  -> {ok, {Location, ?T_MINUS}};
        _   ->
          if $0 =< Ch, Ch =< $9 -> do_number(ScanState, Location, Ch);
             ($A =< Ch andalso Ch =< $Z) orelse
             ($a =< Ch andalso Ch =< $z) orelse
             Ch =:= $. orelse
             Ch =:= $$ orelse
             Ch =:= $_ -> do_symbol(ScanState, Location, [Ch]);
             true -> badchar(ScanState, Ch, "")
           end
      end
  end.

%% Scan after seeing '#'.
do_line_comment(ScanState) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> badchar(ScanState, eof, "in line comment");
    {ok, $\n} -> {ok, {location(ScanState), ?T_NEWLINE}};
    {ok, _Ch} -> do_line_comment(ScanState)
  end.

%% Scan after seeing '/'.
do_slash(ScanState) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    {ok, $*} -> do_c_comment(ScanState, false);
    {ok, Ch} ->
      ungetc(Ch, ScanState),
      badchar(ScanState, Ch, "after /"); % TODO: NYI: T_DIV
    eof ->
      badchar(ScanState, eof, "after /")
  end.

%% Scan after seeing '/* ...'.
do_c_comment(ScanState, PrevWasStar) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> badchar(ScanState, eof, "in /*...*/ comment");
    {ok, $*} -> do_c_comment(ScanState, true);
    {ok, $/} when PrevWasStar -> token(ScanState);
    {ok, _Ch} -> do_c_comment(ScanState, false)
  end.

%% Scan after seeing '"'.
do_string(ScanState, Location, Chars) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> badchar(ScanState, eof, "in string literal");
    {ok, $\n} -> badchar(ScanState, $\n, "in string literal");
    {ok, $\"} -> {ok, {Location, {?T_STRING, lists:reverse(Chars)}}};
    {ok, $\\} ->
      case do_escape(ScanState) of
        {error, _Reason} = Error -> Error;
        {ok, Ch} -> do_string(ScanState, Location, [Ch | Chars])
      end;
    {ok, Ch} -> do_string(ScanState, Location, [Ch | Chars])
  end.

%% Scan after seeing '\' in a string literal.
do_escape(ScanState) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> badchar(ScanState, eof, "in \\ character escape");
    {ok, Ch} ->
      case Ch of
        $n  -> {ok, $\n};
        $t  -> {ok, $\t};
        $f  -> {ok, $\f};
        $r  -> {ok, $\r};
        $b  -> {ok, $\b};
        $\\ -> {ok, $\\};
        $\' -> {ok, $\'};
        $\" -> {ok, $\"};
        _   ->
          if $0 =< Ch, Ch =< $7 -> do_octal_escape(ScanState, Ch - $0, 2);
             true -> badchar(ScanState, Ch, "in \\ character escape")
          end
      end
  end.

do_octal_escape(_ScanState, Val, 0) -> {ok, Val};
do_octal_escape(ScanState, Val, N) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> badchar(ScanState, eof, "in \\ character escape");
    {ok, Ch} ->
      if $0 =< Ch, Ch =< $t -> do_octal_escape(ScanState, Val * 8 + (Ch - $0), N - 1);
         true ->
           case ungetc(Ch, ScanState) of
             {error, _Reason} = Error -> Error;
             ok -> {ok, Val}
           end
       end
  end.

do_symbol(ScanState, Location, Chars) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> do_symbol(Location, lists:reverse(Chars));
    {ok, Ch} ->
      if ($A =< Ch andalso Ch =< $Z) orelse
         ($a =< Ch andalso Ch =< $z) orelse
         ($0 =< Ch andalso Ch =< $9) orelse
         Ch =:= $. orelse
         Ch =:= $$ orelse
         Ch =:= $_ -> do_symbol(ScanState, Location, [Ch | Chars]);
         true ->
           case ungetc(Ch, ScanState) of
             {error, _Reason} = Error -> Error;
             ok -> do_symbol(Location, lists:reverse(Chars))
           end
      end
  end.

do_symbol(Location, Chars) ->
  case Chars of
    [$.] -> {ok, {Location, ?T_DOT}};
    [$. | _] -> {ok, {Location, token:from_symbol(Chars)}};
    _ -> {ok, {Location, {?T_SYMBOL, Chars}}}
  end.

do_number(ScanState, Location, Dig0) ->
  case Dig0 of
    $0 ->
      case fgetc(ScanState) of
        {error, _Reason} = Error -> Error;
        eof -> {ok, {Location, {?T_UINTEGER, Dig0 - $0}}};
        {ok, Ch} ->
          if Ch =:= $x; Ch =:= $X ->
               %% must have hex digit after 0x
               case fgetc(ScanState) of
                 {error, _Reason} = Error -> Error;
                 eof -> badchar(ScanState, eof, "after 0x in number");
                 {ok, Ch} ->
                   case chval(Ch) of
                     ChVal when ChVal < 16 ->
                       do_number(ScanState, Location, _Base = 16, ChVal);
                     _Val -> badchar(ScanState, Ch, "after 0x in number")
                   end
               end;
             true ->
               case ungetc(Ch, ScanState) of
                 {error, _Reason} = Error -> Error;
                 ok -> do_number(ScanState, Location, _Base = 8, _Val = 0)
               end
          end
      end;
    _ -> do_number(ScanState, Location, _Base = 10, _Val = Dig0 - $0)
  end.

do_number(ScanState, Location, Base, Val) ->
  case fgetc(ScanState) of
    {error, _Reason} = Error -> Error;
    eof -> {ok, {Location, {?T_UINTEGER, Val}}};
    {ok, Ch} ->
      case chval(Ch) of
        ChVal when ChVal < Base ->
          do_number(ScanState, Location, Base, Val * Base + ChVal);
        _ChVal when Base =< 10 andalso (Ch =:= $b orelse Ch =:= $f) ->
          {ok, {Location, {?T_LOCAL_LABEL, Val, Ch}}};
        _ChVal ->
          case ungetc(Ch, ScanState) of
            {error, _Reason} = Error -> Error;
            ok -> {ok, {Location, {?T_UINTEGER, Val}}}
          end
      end
  end.

chval(Ch) ->
  if $0 =< Ch, Ch =< $9 -> Ch - $0;
     $A =< Ch, Ch =< $F -> Ch - ($A - 10);
     $a =< Ch, Ch =< $f -> Ch - ($a - 10);
     true -> 255 % signals invalid by being >= any valid base
  end.

badchar(ScanState, Ch, Context) ->
  {ok, {FileName, LineNr}} = location(ScanState),
  {error, {?MODULE, {FileName, LineNr, Ch, Context}}}.

-spec format_error(term()) -> io_lib:chars().
format_error(ungetc) -> "internal error: invalid ungetc";
format_error({FileName, LineNr, Ch, Context}) ->
  io_lib:format("~s line ~p: invalid character '~s' ~s",
                [FileName, LineNr, char_to_string(Ch), Context]).

char_to_string(eof) -> "<EOF>";
char_to_string(Ch) when $\s =< Ch, Ch =< $~ -> [$', Ch, $'];
char_to_string(Ch) ->
  [$', $\\, $0 + ((Ch bsr 6) band 3), $0 + ((Ch bsr 3) band 7), $0 + (Ch band 7), $'].
